[Resubmission] Add a documentation page for DDP communication hooks

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51846 [Gradient Compression] Add register_comm_hook API to DDP communication hooks documentation page
* **#51773 [Resubmission] Add a documentation page for DDP communication hooks**

Resubmission of #51715.

Fixes:

1. Removed "Note [Guidance to Tune ``matrix_approximation_rank`` And ``start_powerSGD_iter``]" in powerSGD_hook.py.
2. Removed the duplicate description of `torch.nn.parallel.DistributedDataParallel.register_comm_hook` in ddp_comm_hooks.rst, because it is already covered by distributed.rst.

Also updated the doc based on the comments from PowerSGD paper author Thijs Vogels .

It seems that `python_doc_test` was flaky. The previous error message was not informative:
https://app.circleci.com/pipelines/github/pytorch/pytorch/270682/workflows/8d186a3c-d682-46bf-b617-ad4eef5991e2/jobs/10739143, and all the warnings did also appear on the master branch.

Rebasing to a new master branch seems to get this fixed:
https://app.circleci.com/pipelines/github/pytorch/pytorch/270696/workflows/1a3adbea-6443-4876-b87b-e17d90d41428/jobs/10740021/steps

Screenshot:
![Screen Shot 2021-02-04 at 10 37 44 PM](https://user-images.githubusercontent.com/8884619/107000422-b2c6be00-673c-11eb-8cba-4202a017b3d7.png)

Differential Revision: [D26272687](https://our.internmc.facebook.com/intern/diff/D26272687/)

Commit:
[Resubmission] Add a documentation page for DDP communication hooks

Resubmission of #51715.

Screenshot:

{F369899792}

Differential Revision: [D26272687](https://our.internmc.facebook.com/intern/diff/D26272687/)

[ghstack-poisoned]

rohan-varma reviewed (commented):
It looks like the doc build is still broken, see:
```
Feb 05 07:35:29 Note [Guidance to Tune ``matrix_approximation_rank`` And ``start_powerSGD_iter``]
Feb 05 07:35:29 ~~~~~~~~~~~~~~~~~~~~~~~~~~
Feb 05 07:35:29 /opt/conda/lib/python3.6/site-packages/torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:docstring of torch.distributed.algorithms.ddp_comm_hooks.powerSGD_hook.PowerSGDState:7: WARNING: Unexpected section title.
Feb 05 07:35:29 
Feb 05 07:35:29 Note [Guidance to Tune ``matrix_approximation_rank`` And ``start_powerSGD_iter``]
```

Commit:
Update on "[Resubmission] Add a documentation page for DDP communication hooks"


Resubmission of #51715.

Increased the length of underline under "Note [Guidance to Tune ``matrix_approximation_rank`` And ``start_powerSGD_iter``]" in powerSGD_hook.py.

Screenshot:
![Screen Shot 2021-02-04 at 10 37 44 PM](https://user-images.githubusercontent.com/8884619/107000422-b2c6be00-673c-11eb-8cba-4202a017b3d7.png)

Differential Revision: [D26272687](https://our.internmc.facebook.com/intern/diff/D26272687/)

[ghstack-poisoned]

Commit:
Update on "[Resubmission] Add a documentation page for DDP communication hooks"


Resubmission of #51715.

Removed "Note [Guidance to Tune ``matrix_approximation_rank`` And ``start_powerSGD_iter``]" in powerSGD_hook.py.

Screenshot:
![Screen Shot 2021-02-04 at 10 37 44 PM](https://user-images.githubusercontent.com/8884619/107000422-b2c6be00-673c-11eb-8cba-4202a017b3d7.png)

Differential Revision: [D26272687](https://our.internmc.facebook.com/intern/diff/D26272687/)

[ghstack-poisoned]

SciPioneer commented:
> It looks like the doc build is still broken, see:
> 
> ```
> Feb 05 07:35:29 Note [Guidance to Tune ``matrix_approximation_rank`` And ``start_powerSGD_iter``]
> Feb 05 07:35:29 ~~~~~~~~~~~~~~~~~~~~~~~~~~
> Feb 05 07:35:29 /opt/conda/lib/python3.6/site-packages/torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:docstring of torch.distributed.algorithms.ddp_comm_hooks.powerSGD_hook.PowerSGDState:7: WARNING: Unexpected section title.
> Feb 05 07:35:29 
> Feb 05 07:35:29 Note [Guidance to Tune ``matrix_approximation_rank`` And ``start_powerSGD_iter``]
> ```

First tried to increase the length of underline, but somehow the update was not picked up by the test.

Now these two lines are removed.

Commit:
Update on "[Resubmission] Add a documentation page for DDP communication hooks"


Resubmission of #51715.

Removed "Note [Guidance to Tune ``matrix_approximation_rank`` And ``start_powerSGD_iter``]" in powerSGD_hook.py.

Screenshot:
![Screen Shot 2021-02-04 at 10 37 44 PM](https://user-images.githubusercontent.com/8884619/107000422-b2c6be00-673c-11eb-8cba-4202a017b3d7.png)

Differential Revision: [D26272687](https://our.internmc.facebook.com/intern/diff/D26272687/)

[ghstack-poisoned]

Commit:
Update on "[Resubmission] Add a documentation page for DDP communication hooks"


Resubmission of #51715.

Removed "Note [Guidance to Tune ``matrix_approximation_rank`` And ``start_powerSGD_iter``]" in powerSGD_hook.py.

Removed the duplicate description of `torch.nn.parallel.DistributedDataParallel.register_comm_hook` in ddp_comm_hooks.rst, because it is already covered by distributed.rst.

Screenshot:
![Screen Shot 2021-02-04 at 10 37 44 PM](https://user-images.githubusercontent.com/8884619/107000422-b2c6be00-673c-11eb-8cba-4202a017b3d7.png)

Differential Revision: [D26272687](https://our.internmc.facebook.com/intern/diff/D26272687/)

[ghstack-poisoned]

tvogels commented on docs/source/ddp_comm_hooks.rst:
```diff
@@ -0,0 +1,68 @@
+    As a biased compressor, PowerSGD requires an extra copy of gradients to
+    enable error feedback for a higher accuracy.
+    This may be infeasible for the use cases that have a memory constraint.
+
```
Suggestion:

```
PowerSGD requires extra memory of the same size as the model's gradients.
It uses this memory for error feedback, to compensate for biased compressed communication.
```

By starting this with the 'extra memory', I think it is more obvious that the warning is about the memory, not about biased compression.

tvogels commented on docs/source/ddp_comm_hooks.rst:
```diff
@@ -0,0 +1,68 @@
+    enable error feedback for a higher accuracy.
+    This may be infeasible for the use cases that have a memory constraint.
+
+.. warning ::
```
Maybe we should add a warning above here about compression rate. I think this should be mentioned as early on as possible in a visible place.

```
.. warning ::
    The compression rate of PowerSGD is determined by the 'matrix approximation rank' specified by the user.
    The lower the rank, the stronger the compression.
    If the rank is too low, you may need more training than normally to reach full model quality.
    The optimal rank depends on the training problem, and we suggest to set its value empirically based on an exponential grid search (1, 2, 4, ...).
```

SciPioneer replied:
Incorporated this into docstrings of `PowerSGDState` in powerSGD_hook.py. The meaning of approximation rank as well as the guidance on the tuning are very important, and I want this also appear in the python file.

tvogels commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -33,22 +33,21 @@ def _orthogonalize(matrix, epsilon=1e-8):
     they are typically turned on for performance.
 
-    Note [Guidance to Tune `matrix_approximation_rank` And `start_powerSGD_iter`]
-    ~~~~~~~~~~~~~~~~~~~~~~~~~~
-    1) To tune `matrix_approximation_rank`, the user can increase it from 1 by factors of 2,
+    1. To tune ``matrix_approximation_rank``, the user can increase it from 1 by factors of 2,
     until a satisfying accuracy can be reached.
-    The increase of `matrix_approximation_rank` can substantially increase the computation costs of the compression.
-    However, the accuracy may not be futher improved beyond a certain `matrix_approximation_rank` value.
-    2) To tune `start_powerSGD_iter`, the user can typically start with 10% of total training steps,
+    The increase of ``matrix_approximation_rank`` can substantially increase the computation costs of the compression.
+    However, the accuracy may not be futher improved beyond a certain ``matrix_approximation_rank`` value.
+
+    2. To tune ``start_powerSGD_iter``, the user can typically start with 10% of total training steps,
     and increase it until a satisfying accuracy can be reached.
     Deferrring PowerSGD can effectively improve the accuracy,
-    even a relatively small `matrix_approximation_rank` is used.
+    even a relatively small ``matrix_approximation_rank`` is used.
     This is because that, the beginning of training phase is usually very sensitive to inaccurate gradients,
     and compressing gradients too early may make the training quickly take a suboptimal trajectory,
     which can result in an irrecoverable impact on the accuracy.
```
Suggested rephrasing:
```
Stores both the algorithm's hyperparameters and the internal state for all the gradients during the training.

The settings ``matrix_approximation_rank`` and ``start_powerSGD_iter`` should be tuned by the user. 
- To tune ``matrix_approximation_rank``, start from 1 and increase by factors of 2 until a satisfactory accuracy is reached. 
A too low ``matrix_approximation_rank`` can yield loss in accuracy, while a high ``matrix_approximation_rank`` yields less communication savings and slower compression. 
- To tune ``start_powerSGD_iter`, we suggest to start with 10% of the total training steps, and increasing it until satisfactory accuracy is reached. Because the beginning of training is the most sensitive to accurate updates, starting to compress communication only after some iterations improves accuracy, even when a relatively small ``matrix_approximation_rank`` is used.

We suggest to keep the binary hyperparameters ``use_error_feedback`` and ``warm_start`` on.
```

tvogels commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -162,38 +161,44 @@ def maybe_increase_iter(self, bucket):
 
 def powerSGD_hook(state: PowerSGDState, bucket) -> torch.futures.Future:
-    """
+    r"""
```
I would drop 'original' in this line.

tvogels commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -162,38 +161,44 @@ def maybe_increase_iter(self, bucket):
-        3.5) Computes each Q in Qs, which is approximately equal to M^TP;
-        3.6) Allreduces Qs as a batch;
-        3.7) Computes each M among all the high-rank tensors, which is approximately equal to PQ^T.
-
-    Note that this communication hook enforces vanilla allreduce for the first `state.start_powerSGD_iter` iterations.
+
+        3.2. Computes each P in Ps, which is equal to MQ;
+
+        3.3. Allreduces Ps as a batch;
+
+        3.4. Orthogonalizes each P in Ps;
+
+        3.5. Computes each Q in Qs, which is approximately equal to M^TP;
+
+        3.6. Allreduces Qs as a batch;
+
+        3.7. Computes each M among all the high-rank tensors, which is approximately equal to PQ^T.
```
I feel like the #.# steps here are maybe too detailed for the end-user documentation.
Maybe you could refer the reader to the paper for details on the algorithm.
The content here also duplicates inline comments, and could lead to inconsistencies.

Maybe something higher level could be
```
2. Gradient tensors are reshaped into a matrix M, where the rows and columns correspond to input- and output dimensions of the layer. 
3. After adding accumulated compression errors from before, this matrix is approximated as low-rank matrix that can be cheaply averaged between workers.
```

SciPioneer replied:
A bit more details are fine, because the pseudocode in the paper cannot cover all the details, such as handling rank-1 tensors, and Ps and Qs are allreduced in batches, etc.

tvogels replied:
Ok, makes sense. I guess there are two targets for this doc: people who want to work with this code and people who just want to use it. To satisfy both, maybe you could should start with a simple and short overview of what happens (along the lines of the suggestion here or even shorter), and then provide details afterwards?

tvogels commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -162,38 +161,44 @@ def maybe_increase_iter(self, bucket):
-    3) Handles high-rank tensors by PowerSGD compression:
-        3.1) For each high-rank tensor M, creates two low-rank tensors P and Q for decomposing M,
+
+    1. Views the input flattened 1D gradient tensor as two groups of per-parameter tensors: high-rank tensors and vector-like rank-1 tensors (for biases).
+
+    2. Handles rank-1 tensors by allreducing them without compression:
+
+        2.1. Allocate contiguous memory for those rank-1 tensors, and allreduces all the rank-1 tensors as a batch, without compression;
+
+        2.2. Copies the individual rank-1 tensors from the contiguous memory back to the input tensor.
```
Maybe point 1 and 2 could be summarized as simply:

```
1. Rank-1 gradients (corresponding to bias parameters) are not compressed and averaged between workers with regular all-reduce communication.
```

I feel like the allocation of contiguous memory is an implementation detail.

SciPioneer replied:
Prefer keeping the details here. This was once asked by a code reviewer, as the pseudocode in the paper does not  expose such details.

There is actually another API `allreduce_coalesced` that can potentially allreduce these rank-1 tensors without explicit packing. Although this API is not supported on NCCL yet, the details here can remind that here can be a good use case.

tvogels commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -162,38 +161,44 @@ def maybe_increase_iter(self, bucket):
+
+        3.7. Computes each M among all the high-rank tensors, which is approximately equal to PQ^T.
+
+    Note that this communication hook enforces vanilla allreduce for the first ``state.start_powerSGD_iter`` iterations.
```
Minor suggestion:

```
This communication hook enforces vanilla allreduce for the first ``state.start_powerSGD_iter`` iterations.
This not only gives the user more control over the tradeoff between speedup and accuracy,
but also help abstract away some complexity of the internal optimization of DDP for future communication hook developers.
```

tvogels commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -412,40 +421,44 @@ def decompress(fut):
-    This variant is faster than `powerSGD_hook` that runs layer-wise gradient compression,
-    but it usually results in a much lower accuracy, unless `matrix_approximation_rank` in the state is 1.
-    Increasing `matrix_approximation_rank` may not necessarily increase the accuracy,
+
```
Again, I think all the steps are perhaps too much detail. I think the first paragraph should clearly state the difference between the standard algorithm.
```
This DDP communication hook implements a simplified PowerSGD gradient compression
    algorithm described in https://arxiv.org/abs/1905.13727.
It doesn't compress the gradients parameter-by-parameter, but instead compresses a flattened representation of all model parameters. This is **comutationally lightweight** compared to the original algorithm, but usually results in **much lower accuracy**.
```

tvogels reviewed (changes requested):
Hi @SciPioneer, I suggested some rephrasing inline. Happy to discuss.

SciPioneer commented on docs/source/ddp_comm_hooks.rst:
```diff
@@ -0,0 +1,68 @@
+PowerSGD Communication Hook
+---------------------------
+
+PowerSGD communication hook is a **stateful** hook used for gradient
+compression, and the user needs to provide a state defined as below.
+The performance is `on par with <https://observablehq.com/@tvogels/powersgd-benchmark>`_
+the implementation in the original `paper <https://arxiv.org/abs/1905.13727>`_.
```
Updated with more

Also moved the reference of benchmark to Acknowledgements section.

Commit:
Update on "[Resubmission] Add a documentation page for DDP communication hooks"


Resubmission of #51715.

Removed "Note [Guidance to Tune ``matrix_approximation_rank`` And ``start_powerSGD_iter``]" in powerSGD_hook.py.

Removed the duplicate description of `torch.nn.parallel.DistributedDataParallel.register_comm_hook` in ddp_comm_hooks.rst, because it is already covered by distributed.rst.

Screenshot:
![Screen Shot 2021-02-04 at 10 37 44 PM](https://user-images.githubusercontent.com/8884619/107000422-b2c6be00-673c-11eb-8cba-4202a017b3d7.png)

Differential Revision: [D26272687](https://our.internmc.facebook.com/intern/diff/D26272687/)

[ghstack-poisoned]

SciPioneer commented:
> Hi @SciPioneer, I suggested some rephrasing inline. Happy to discuss.

Thanks for all the comments!

tvogels reviewed (approved):
Looks great!

Commit:
Update on "[Resubmission] Add a documentation page for DDP communication hooks"


Resubmission of #51715.

Fixes:

1. Removed "Note [Guidance to Tune ``matrix_approximation_rank`` And ``start_powerSGD_iter``]" in powerSGD_hook.py.
2. Removed the duplicate description of `torch.nn.parallel.DistributedDataParallel.register_comm_hook` in ddp_comm_hooks.rst, because it is already covered by distributed.rst.

Also updated the doc based on the comments from PowerSGD paper author Thijs Vogels .

Screenshot:
![Screen Shot 2021-02-04 at 10 37 44 PM](https://user-images.githubusercontent.com/8884619/107000422-b2c6be00-673c-11eb-8cba-4202a017b3d7.png)

Differential Revision: [D26272687](https://our.internmc.facebook.com/intern/diff/D26272687/)

[ghstack-poisoned]

Commit:
Update on "[Resubmission] Add a documentation page for DDP communication hooks"


Resubmission of #51715.

Fixes:

1. Removed "Note [Guidance to Tune ``matrix_approximation_rank`` And ``start_powerSGD_iter``]" in powerSGD_hook.py.
2. Removed the duplicate description of `torch.nn.parallel.DistributedDataParallel.register_comm_hook` in ddp_comm_hooks.rst, because it is already covered by distributed.rst.

Also updated the doc based on the comments from PowerSGD paper author Thijs Vogels .

Screenshot:
![Screen Shot 2021-02-04 at 10 37 44 PM](https://user-images.githubusercontent.com/8884619/107000422-b2c6be00-673c-11eb-8cba-4202a017b3d7.png)

Differential Revision: [D26272687](https://our.internmc.facebook.com/intern/diff/D26272687/)

[ghstack-poisoned]
